File: java/src/main/java/com/github/algo/dp/SubarraySumEqualsK.java
package com.github.algo.dp;

import java.util.HashMap;
import java.util.Map;

public class SubarraySumEqualsK {

    public int subarraySum(int[] nums, int k) {
        Map<Integer, Integer> cache = new HashMap<>();
        cache.put(0, 1);
        int currentSum = 0;
        int count = 0;
        for (int i = 0; i < nums.length; i++) {
            currentSum += nums[i];
            if (cache.containsKey(currentSum - k)) {
                count += cache.get(currentSum - k);
            }
            cache.put(currentSum, cache.getOrDefault(currentSum, 0) + 1);
        }
        return count;
    }

}

File: python/main/arrays/SubarraySumEqualsK.py
from typing import List


class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        cache = {0: 1}
        currentSum = 0
        count = 0
        for n in nums:
            currentSum += n
            if currentSum - k in cache:
                count += cache[currentSum - k]
            if currentSum not in cache:
                cache[currentSum] = 0
            cache[currentSum] = cache[currentSum] + 1
        return count


if __name__ == "__main__":
    solution = Solution()
    isPossible = solution.subarraySum([1, 1, 1], 2)
    print(isPossible)
